include/envoy/http/header_map.h

```diff
@@ -102,6 +102,31 @@ class HeaderString {
    */
   bool find(const char* str) const { return strstr(c_str(), str); }
 
+  /**
+   * HeaderString is in token list form, each token separated by commas or whitespace,
+   * see https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.1 for more information,
+   * header field value's case sensitivity depends on each header.
+   * @return whether contains token in case insensitive manner.
+   */
+  bool caseInsensitiveContains(const char* token) const {
+    // Avoid dead loop if token argument is empty.
+    const int n = strlen(token);
+    if (n == 0) {
+      return true;
+    }
+
+    // Find token substring, skip if it's partial of other token.
+    const char* tokens = c_str();
+    for (const char* p = tokens; (p = strcasestr(p, token)); p += n) {
+      if ((p == tokens || *(p - 1) == ' ' || *(p - 1) == ',') &&
+          (*(p + n) == '\0' || *(p + n) == ' ' || *(p + n) == ',')) {
```

Comment:
Don't you need to check here that the length of the remaining header is >= n before you do this? I think this code can overrun the header. Please make sure you add more test cases to cover small headers.

Reply:
Because it advance the pointer p only if found the token, there is no overrun. I will add more test cases cover small headers.

Reply:
Ah I see. Sorry got it. Yup you are right. Please just add some small header tests and we are good.